### PR TITLE
Add managed disk option to azure driver

### PR DESCRIPTION
**Pull request by rmweir**
Problem:
rancher does not currently support using managed disk but offer options that require it

Solution:
Support managed disks. This requires that rancher can create managed availability sets, and managed VMs.

Issue:
https://github.com/rancher/rancher/issues/15788

UI Issue:
https://github.com/rancher/rancher/issues/21533

### CODE REVIEW

File: drivers/azure/azureutil/azureutil.go [hunk 1]
```diff
@@ -563,41 +564,105 @@ func (a AzureClient) CreateVirtualMachine(resourceGroup, name, location, size, a
...
+					OsDisk: a.GetOSDisk(name, osDiskBlobURL, isManaged, storageType),
 				},
 			},
 		}, nil)
 	return err
 }
 
+// GetOSDisk creates and returns pointer to a disk that is configured for either managed or unmanaged disks depending
+// on setting.
+func (a AzureClient) GetOSDisk(name string, osDiskBlobURL string, isManaged bool, storageType string) *compute.OSDisk {
```
Comment (luthermonson) [inline drivers/azure/azureutil/azureutil.go]: never use a might as well kill it and make getOSDisk() instead

**Comment by luthermonson** [comment]
lgtm

**Comment by luthermonson** [comment]
disksize update lgtm too

**Comment by luthermonson** [comment]
squash lgtm, gtg!

File: drivers/azure/azure.go [hunk 2]
```diff
@@ -360,7 +395,7 @@ func (d *Driver) Create() error {
 	if err := c.CreateResourceGroup(d.ResourceGroup, d.Location); err != nil {
 		return err
 	}
-	if err := c.CreateAvailabilitySetIfNotExists(d.ctx, d.ResourceGroup, d.AvailabilitySet, d.Location); err != nil {
+	if err := c.CreateAvailabilitySetIfNotExists(d.ctx, d.ResourceGroup, d.AvailabilitySet, d.Location, d.ManagedDisks, int32(d.FaultCount), int32(d.UpdateCount)); err != nil {
```
Comment (daxmc99) [inline drivers/azure/azure.go]: Possible information loss here int32 <- from int

File: drivers/azure/azureutil/azureutil.go [hunk 3]
```diff
@@ -591,13 +611,59 @@ func (a AzureClient) GetAvailabilitySet(resourceGroup, name string) (compute.Ava
 	return a.availabilitySetsClient().Get(resourceGroup, name)
 }
 
-func (a AzureClient) CreateAvailabilitySetIfNotExists(ctx *DeploymentContext, resourceGroup, name, location string) error {
+// CreateAvailabilitySetIfNotExists checks that managed disk option match availability set if it already exists. If the
+// availability set does not already exists than it is created with configured parameters.
+func (a AzureClient) CreateAvailabilitySetIfNotExists(ctx *DeploymentContext, resourceGroup, name, location string, isManaged bool, faultCount int32, updateCount int32) error { // , faultCount int32, updateCount int32) error {
```
Comment (daxmc99) [inline drivers/azure/azureutil/azureutil.go]: Purpose of comment `// , faultCount int32, updateCount int32) error {` on this line?

**Comment by daxmc99** [comment]
LGTM :)

**Comment by dramich** [comment]
lgtm

**Pull request merged**